[DOC] Search tutorial (#775)
seqan/seqan3: doc/tutorial/search/index.md

# Indexing and searching with SeqAn3 {#tutorial_index_search}

<b>Learning Objective:</b><br>
In this tutorial you will learn how to construct an index and conduct searches.

\tutorial_head{Moderate, 60 Minutes, Alignment (only for last assignment), [FM-Index paper](https://doi.org/10.1109/SFCS.2000.892127)<br>[FM-Index on Wikipedia](https://en.wikipedia.org/wiki/FM-index)<br>[Bi-FM-Index paper](https://doi.org/10.1109/BIBM.2009.42)<br>[Search Schemes paper](http://doi.org/10.1101/301085)}

[TOC]

<br><br>

# Introduction

Exact and approximate string matching is a common problem in bioinformatics, e.g. in read mapping.

> This tutorial will show you how to use the seqan3::fm_index and seqan3::bi_fm_index to create indices and how to search them efficiently using seqan3::search.

Reviewer: not resolved?

> This tutorial will show you how to use the seqan3::fm_index and seqan3::bi_fm_index to create indices and how to search them efficiently using seqan3::search.

Author: What do you mean by it? Breaking the line after 120?

> This tutorial will show you how to use the seqan3::fm_index and seqan3::bi_fm_index to create indices and how to search them efficiently using seqan3::search.

Reviewer 2: yes

### Bi_FM-Index
In contrast to the FM-Index, the Bidirectional Index (Bi-FM-Index) allows searches to be conducted in both directions, i.e. from right to left and left to right. Being able to start at any position of the query in a search improves the search performance drastically when allowing multiple errors. You can find more information on Bi-FM-Indicies [here](https://doi.org/10.1109/BIBM.2009.42).<br>
However, it also needs to store additional information to be bidirectional. You can think of the Bi-FM-Index as two FM-Indices, one for each direction.<br>
When searching, both FM-Indices need to be kept in sync, which applies an addtional overhead. Normally this would degredate performance if the search is straightforward on a single FM-Index. In SeqAn3, a search using a bidrectional FM-Index is automatically simplified to a unidirectional search on one of its FM-Indicies whenever this is the case.

> 

Reviewer: I don't know. Have you adapted it? Do you have time to do it before tomorrow?

> 

Author: I think I changed it?

### Query
A query is the data you want to search for in the reference, e.g. a read or an amino acid sequence.
### Index
An index is a data structure usually built over the reference that allows fast searches.
### FM-Index
The full-text minute index is a static data structure that allows for fast queries in the indexed text and is a state-of-the-art index. You can find more information on FM-Indicies [in the original publication](https://doi.org/10.1109/SFCS.2000.892127) and [on Wikipedia](https://en.wikipedia.org/wiki/FM-index).
### Bi-FM-Index
In contrast to the FM-Index, the Bidirectional Index (Bi-FM-Index) allows searches to be conducted in both directions, i.e. from right to left and left to right. Being able to start at any position of the query in a search improves the search performance drastically when allowing multiple errors. You can find more information on Bi-FM-Indicies [here](https://doi.org/10.1109/BIBM.2009.42).<br>
However, it also needs to store additional information to be bidirectional. You can think of the Bi-FM-Index as two FM-Indices, one for each direction, hence you need about 2 times more space for the Bi-FM-Index.<br>

## Example

> Both FM- and Bi-FM-Indices internally use the implementation of the [Succinct Data Structure Library](https://github.com/simongog/sdsl-lite).<br>

Reviewer: I would not mention this if people do not need to interface with the sdsl to work with our code.

Reviewer: This is probably documented in the API already and links to the API exists as well.

\endcode
Create an fm_index over the reference, store the index and load the index into a new fm_index object.
Print whether the store and load operations were successful.
\endassignment

\solution
\snippet doc/tutorial/search/solution1.cpp solution
**Expected output:**
~~~console
Index stored successfully.
Index loaded successfully.
~~~
\endsolution

# Search

Using an index, we can now conduct searches for a given query. In this part we will learn, how to search exactly, allow substitutions and indels, and how to configure what kind of results we want, e.g. all results vs. only the best result.

## Terminology
### Search scheme
A search scheme is the manner of how a search is conducted, e.g. trying to match each position in the reference to the query.
### Optimal Search Scheme
The scheme that requires the least amount of queries in the index is called an *optimal search scheme*.<br>
SeqAn3 provides *optimal* search schemes for up to **4** errors for Bi-FM-Indices and provides sub-optimal, but still efficient search schemes otherwise.<br>
You can read more about search schemes in [this paper](http://doi.org/10.1101/301085).

> \note The choice of the search scheme is done internally you don't have to worry about it. You can only choose between the FM and Bi-FM-Index but when using >= 2 errors it is advisable to use a Bi-FM-Index since searches will be faster.<br>

Reviewer: Please do not mention things in the tutorial that readers don't have to worry about? Otherwise why would you bother explaining to them what it is?

Reviewer: Just say that for more than 2 errors using the Bi-FM-Index is more efficient.

## Searching for exact matches

We can search for all exact matches using seqan3::search:

\include doc/tutorial/search/basic_search.cpp

When using a single text, the search will return a std::vector of the begin positions of matches in the reference.

> For text collections, a tuple where the first element represents the index of the text in the collection and the second element represents the position within the text.<br>

Reviewer: ```suggestion
For text collections, a std::vector over tuples where the first element represents the index of the text in the collection and the second element represents the position within the text.<br>
```

There are 3 hits.
The positions are [1,41,77]
===== Running on a text collection =====
There are 3 hits.
The positions are [(0,1),(1,9),(2,16)]
~~~
\endsolution

## Searching for approximate matches

If you want to allow for errors in your query, you need to configure the approximate search with a seqan3::search_cfg::max_error or seqan3::search_cfg::max_error_rate object. The is constructed with four parameters:

With absolute numbers (seqan3::search_cfg::max_error):

- seqan3::search_cfg::total: Maximum number of total errors
- seqan3::search_cfg::substitution: Maximum number of substitutions
- seqan3::search_cfg::insertion: Maximum number of insertions
- seqan3::search_cfg::deletion: Maximum number of deletions

Use seqan3::search_cfg::max_error_rate with the same elements to define error rates \f$\in[0,1]\f$.

To search for either 1 insertion or 1 deletion, you can write:

\snippet doc/tutorial/search/small_snippets.cpp error_search

Note that the error amount/rates do not have to sum up to the total of errors allowed:
\snippet doc/tutorial/search/small_snippets.cpp error_sum

Allows 2 errors which can be any combination of 2 substituions, 1 insertion and 1 deletion.

> Defining only the total will set all error types to the total.<br>

Reviewer: ```suggestion
Defining only the total will set all error types to this value.<br>
```

> Defining any of the error types, but no total, will set the total to the sum of all error types.

Reviewer: And what happens if you define both and they don't match?

In this tutorial you will learn how to construct an index and conduct searches.

\tutorial_head{Moderate, 60 Minutes, Alignment (only for last assignment), [FM-Index paper](https://doi.org/10.1109/SFCS.2000.892127)<br>[FM-Index on Wikipedia](https://en.wikipedia.org/wiki/FM-index)<br>[Bi-FM-Index paper](https://doi.org/10.1109/BIBM.2009.42)<br>[Search Schemes paper](http://doi.org/10.1101/301085)}

[TOC]

<br><br>

# Introduction

Exact and approximate string matching is a common problem in bioinformatics, e.g. in read mapping.
This tutorial will show you how to use the seqan3::fm_index and seqan3::bi_fm_index to create indices and how to search them efficiently using seqan3::search.

## Capabilities

With this module you can:
* Create, store and load (Bi)-FM-Indices
* Search for exact matches
* Search for approximate matches (allowing substitutions and indels)

The results of the search can be passed on to other module, e.g. to create an alignment.

## Terminology
### Reference
A reference is the data you want to search in, e.g. a genome or a text.
### Query
A query is the data you want to search for in the reference, e.g. a read or a substring.
### Index

> An index is a data structure usually built over the reference that allows fast searches.

Author: Mhm, but you could index the reads and then scan the genome (doesn't RazerS 3 do this?).

> An index is a data structure usually built over the reference that allows fast searches.

Reviewer: You should only document what can be done at the moment. I think once we ported the swift-filter one can add this other use case. Right now searching the reference inside of the reads by means of a FM-index does not make a lot of sense in my opinion.

## Example

Both FM- and Bi-FM-Indices internally use the implementation of the [Succinct Data Structure Library](https://github.com/simongog/sdsl-lite).<br>
Conctructing a (Bi-)FM-Index is very simple as the template parameters are automatically deduced:

\include doc/tutorial/search/basic_index.cpp

You can also index text collections:

\snippet doc/tutorial/search/small_snippets.cpp text_collection

The indices can also be stored and loaded from disk.
Both seqan3::fm_index::store and seqan3::fm_index::load return a `bool` indicating whether the operation was successful.

\snippet doc/tutorial/search/small_snippets.cpp store

\snippet doc/tutorial/search/small_snippets.cpp load

Note that you have to templatise the `fm_index` for default construction. In contrast to the construction via a given `text`, the template cannot be deduced from the type of `text` by the compiler.

\assignment{Assignment 1}
You are given the text
\code

> dna4_vector text{"CGCTGTCTGAAGGATGAGTGTCAGCCAGTGTAACCCGATGAGCTACCCAGTAGTCGAACTGGGCCAGACAACCCGGCGCTAATGCACTCA"_dna4};

Author: I can, but it also adds the indentation

> dna4_vector text{"CGCTGTCTGAAGGATGAGTGTCAGCCAGTGTAACCCGATGAGCTACCCAGTAGTCGAACTGGGCCAGACAACCCGGCGCTAATGCACTCA"_dna4};

Reviewer: But I wouldn't worry about that.

Index loaded successfully.
~~~
\endsolution

# Search

Using an index, we can now conduct searches for a given query. In this part we will learn, how to search exactly, allow substitutions and indels, and how to configure what kind of results we want, e.g. all results vs. only the best result.

## Terminology
### Search scheme
A search scheme is the manner of how a search is conducted. Depending on what kind of errors are allowed, there is an optimal way of search. For example, for 1 allowed substitution the trivial approach is to search for all possible permutations with one substitution. It is more efficient to divide the query into two blocks, search one block exactly and allow the error in the other block. Even though this has to be done twice (error in block 1 and error in block 2), this requires less lookups in the index than the trivial approach.
### Optimal Search Scheme
The scheme that requires the least amount of queries in the index is called an *optimal search scheme* and determining it is NP-hard.<br>
SeqAn3 provides *optimal* search schemes for up to **4** errors for Bi-FM-Indices and provides sub-optimal, but still efficient search schemes otherwise.<br>

> The choice of the search scheme is done internally, but when using >= 2 errors it is advisable to construct a Bi-FM-Index since searches will be faster.<br>

Reviewer 3: ```suggestion

\note The choice of the search scheme is done internally you don't have to worry about it. You can only choose between the FM and Bi-FM-Index but when using >= 2 errors it is advisable to use a Bi-FM-Index since searches will be faster.<br>
```

> You can read more about search schemes in [this paper](http://doi.org/10.1101/301085).

Reviewer 3: put this line before the sentence/note above.

> ## Example

Reviewer 3: ```suggestion
## Searching for exact matches
```

We can search for all exact matches using seqan3::search:

\include doc/tutorial/search/basic_search.cpp

When using a single text, the search will return a std::vector of the begin positions of matches in the reference.
For text collections, a tuple where the first element represents the index of the text in the collection and the second element represents the position within the text.<br>
You can also pass multiple queries at the same time:

\snippet doc/tutorial/search/small_snippets.cpp multiple_queries

The returned result will be a vector of individual results. Each element refers to the respective query.

\assignment{Assignment 2}
Search for all exact occurrences of `GCT` in the text from assignment 1.<br>
Print the number of hits and the positions in a ascending ordering.<br>
Do the same for the following text collection:
\code

> std::vector<dna4_vector> text{"CGCTGTCTGAAGGATGAGTGTCAGCCAGTGTA"_dna4,

Reviewer 3: ssame as above. use the snippet

                              "ACCCGATGAGCTACCCAGTAGTCGAACTG"_dna4,
                              "GGCCAGACAACCCGGCGCTAATGCACTCA"_dna4};
\endcode
\endassignment

\solution
\include doc/tutorial/search/solution2.cpp
**Expected output:**
~~~console
=====   Running on a single text   =====
There are 3 hits.
The positions are [1,41,77]
===== Running on a text collection =====
There are 3 hits.
The positions are [(0,1),(1,9),(2,16)]
~~~
\endsolution

> ## Configuration

Reviewer 3: ```suggestion
## Searching for approximate matches
```

> We offer the following configurations for allowing errors:

Reviewer 3: Something along the lines
```suggestion
If you want to allow for errors in your query, you need to configure the approximate search with a seqan3::search_cfg::max_error or seqan3::search_cfg::max_error_rate object. The is constructed with four parameters:
```
Question: are all  combinations of numbers possible?

With absolute numbers (seqan3::search_cfg::max_error):

- seqan3::search_cfg::total: Maximum number of total errors
- seqan3::search_cfg::substitution: Maximum number of substitutions
- seqan3::search_cfg::insertion: Maximum number of insertions
- seqan3::search_cfg::deletion: Maximum number of deletions

Use seqan3::search_cfg::max_error_rate with the same elements to define error rates \f$\in[0,1]\f$.

To search for either 1 insertion or 1 deletion, you can write:

\snippet doc/tutorial/search/small_snippets.cpp error_search

> Note that the error type do not have to sum up to the total of errors allowed:

Reviewer 3: ```suggestion
Note that the error rates do not have to sum up to the total of errors allowed:
```
?

~~~cpp

> configuration const cfg = max_error{total{2}, substitution{2}, insertion{1}, deletion{1}};

Reviewer 3: snippet!

~~~
Allows 2 errors which can be any combination of 2 substituions, 1 insertion and 1 deletion.

Defining only the total will set all error types to the total.<br>
Defining any of the error types, but no total, will set the total to the sum of all error types.

\assignment{Assignment 3}
Search for all occurrences of `GCT` in the text from assignment 1.<br>
Allow up to 1 substitution and print the matching substring of the reference.
\endassignment

\solution
\include doc/tutorial/search/solution3.cpp
**Expected output:**
~~~console
There are 12 hits.
At position 1: GCT
At position 5: TCT
At position 12: GAT
At position 23: GCC
At position 36: GAT
At position 41: GCT
At position 57: ACT
At position 62: GCC
At position 75: GCG
At position 77: GCT
At position 83: GCA
At position 85: ACT
~~~
\endsolution

> ### Search modes

Reviewer 3: ```suggestion
## Search modes
```

> Additionally, you can define what kind of hits should be reported:

Reviewer 3: ```suggestion
Besides the error configuration, you can define what kind of hits should be reported:
```

> - seqan3::search_cfg::all: Report all hits that satisfy the requirement.

Reviewer 3: ```suggestion
- seqan3::search_cfg::all: Report all hits that satisfy the (approximate) search.
```
requirements is very general

> - seqan3::search_cfg::best: Report the best hit, i.e. the hit with the lowest edit distance.

Reviewer 3: what happens if there are several best hits? which one is chosen? the first?
add a comment

- seqan3::search_cfg::all_best: Report all hits with the lowest edit distance.
- seqan3::search_cfg::strata: best+x mode. Report all hits within the x-neighbourhood of the best hit.

The mode is appended to the error configuration by piping:
~~~cpp

> configuration const cfg = max_error{total{1}, substitution{0}, insertion{1}, deletion{1}} | mode::best;

Reviewer 3: sniiiiiipppet :D

~~~

The strata mode needs an additional parameter:
~~~cpp
configuration const cfg = max_error{total{2}, substitution{0}, insertion{1}, deletion{1}} | mode::strata{2};
~~~

If the best hit had an edit distance of 1, the strata mode would report all hits with up to an edit distance of 3. Since in this example the total error number is given as 2, all hits with 1 or 2 errors would be reported.

\assignment{Assignment 4}
Search for all occurrences of `GCT` in the text from assignment 1.<br>

> Allow up to 1 error of any type and print the number of hits for each search mode (use `mode::strata{1}`).

Reviewer 3: ```suggestion
Allow up to 1 error of any type and print the number of hits for each search mode.
\hint
Use `mode::strata{1}`.
\endhint
```

Author: It's not really a hint since they should use the 1-strata for the strata search mode

\endassignment

\solution
\include doc/tutorial/search/solution4.cpp
**Expected output:**
~~~console
Searching all hits
There are 25 hits.
Searching all best hits
There are 3 hits.
Searching best hit
There is 1 hit.
Searching all hits in the 1-stratum
There are 25 hits.
~~~
\endsolution

> # Exercise

Reviewer 3: ```suggestion
## One last exercise
```

\endcode
Create a seqan3::fm_index over the reference, store the index and load the index into a new seqan3::fm_index object.
Print whether the store and load operations were successful.
\endassignment

\solution
\snippet doc/tutorial/search/search_solution1.cpp solution
**Expected output:**
```console
Index stored successfully.
Index loaded successfully.
```
\endsolution

# Search

Using an index, we can now conduct searches for a given query. In this part we will learn how to search exactly, allow substitutions and indels, and how to configure what kind of results we want, e.g. all results vs. only the best result.

## Terminology
### Search Scheme
A search scheme is the manner of how a search is conducted, e.g. trying to match each position in the reference to the query.
### Optimal Search Scheme
The scheme that requires the least amount of queries in the index is called an *optimal search scheme*.

SeqAn3 provides *optimal* search schemes for up to **4** errors for Bi-FM-Indices and provides sub-optimal but still efficient search schemes otherwise.
You can read more about search schemes in [this paper](http://doi.org/10.1101/301085).

> 

Author: Mhm, it's not that important since it happens in the background.

But it's maybe important to know that for more than 4 errors the search get's way slower since the scheme is not optimal anymore.

And it's something fancy we have, so we might as well advertise it? 😁

> 

Reviewer 2: > But it's maybe important to know that for more than 4 errors the search get's way slower since the scheme is not optimal anymore.

Who searches with 4 error anyway? :wink: 

> And it's something fancy we have, so we might as well advertise it?

Except that our readers don't understand the fanciness and just think "seqan is complicated" :expressionless:

> 

Reviewer 4: I agree with Hannes. Hopefully we will soon have a generic efficient solution for >4 errors :)

I implemented 01*0 seeds for my thesis formulated as search schemes for which I could make a PR after I finished work on the SDSL.

> 

Reviewer 2: To be more precise: If I read that something is optimal until 4 errors and not optimal after that, it sounds like it will misbehave like return worse/bad results or miss something after four errors. I would really not write this in the tutorial. It has its place in the API docs, but not here.
